runtime/sema/type_tags.go

```diff
@@ -650,6 +653,120 @@ func getSuperTypeOfDerivedTypes(types []Type) Type {
 	return prevType
 }
 
+func commonSuperTypeOfVariableSizedArrays(types []Type) Type {
+	// We reach here if all types are variable-sized arrays.
+	// Therefore, decide the common supertype based on the element types.
+
+	elementTypes := make([]Type, 0)
+
+	for _, typ := range types {
+		// 'Never' type doesn't affect the supertype.
+		// Hence, ignore them
+		if typ == NeverType {
+			continue
+		}
+
+		arrayType, ok := typ.(*VariableSizedType)
+		if !ok {
+			panic(fmt.Errorf("expected variabled-sized array type, found %s", typ))
+		}
+
+		elementTypes = append(elementTypes, arrayType.ElementType(false))
+	}
+
+	elementSuperType := LeastCommonSuperType(elementTypes...)
+
+	if elementSuperType == InvalidType {
+		return InvalidType
+	}
+
+	return &VariableSizedType{
+		Type: elementSuperType,
+	}
+}
+
+func commonSuperTypeOfConstantSizedArrays(types []Type) Type {
+	// We reach here if all types are constant-sized arrays.
+	// Therefore, decide the common supertype based on the element types.
+
+	elementTypes := make([]Type, 0)
+	var prevType *ConstantSizedType
+
+	for _, typ := range types {
+		// 'Never' type doesn't affect the supertype.
+		// Hence, ignore them
+		if typ == NeverType {
+			continue
+		}
+
+		arrayType, ok := typ.(*ConstantSizedType)
+		if !ok {
+			panic(fmt.Errorf("expected constnant-sized array type, found %s", typ))
+		}
+
+		elementTypes = append(elementTypes, arrayType.ElementType(false))
+
+		if prevType == nil {
+			prevType = arrayType
+			continue
+		}
+
+		// Arrays with different sizes are not covariant
+		if arrayType.Size != prevType.Size {
+			return commonSuperTypeOfHeterogeneousTypes(types)
+		}
+	}
+
+	elementSuperType := LeastCommonSuperType(elementTypes...)
+
+	if elementSuperType == InvalidType {
+		return InvalidType
+	}
+
+	return &ConstantSizedType{
+		Type: elementSuperType,
+		Size: prevType.Size,
+	}
+}
+
+func commonSuperTypeOfDictionaries(types []Type) Type {
+	// We reach here if all types are dictionary types.
+	// Therefore, decide the common supertype based on the key types and value types.
+
+	keyTypes := make([]Type, 0)
+	valueTypes := make([]Type, 0)
+
+	for _, typ := range types {
+		// 'Never' type doesn't affect the supertype.
+		// Hence, ignore them
+		if typ == NeverType {
+			continue
+		}
+
+		dictionaryType, ok := typ.(*DictionaryType)
+		if !ok {
+			panic(fmt.Errorf("expected dictionary type, found %s", typ))
+		}
+
+		valueTypes = append(valueTypes, dictionaryType.ValueType)
+		keyTypes = append(keyTypes, dictionaryType.KeyType)
+	}
+
+	keySuperType := LeastCommonSuperType(keyTypes...)
+	valueSuperType := LeastCommonSuperType(valueTypes...)
+
+	if keySuperType == InvalidType ||
+		valueSuperType == InvalidType ||
+		!IsValidDictionaryKeyType(keySuperType) {
```

Comment:
I'm not sure we want to produce an invalid type when the dictionary key is invalid. Consider a case where we have two dictionaries: `{String: Int}` and `{Int: String}`. In this case, both `keySuperType` and `valueSuperType` are going to be `AnyStruct`, but I don't think this means the inferred type should itself be invalid; we probably want to use the appropriate `AnyStruct` or `AnyResource` in this case.

Reply:
Currently, the checker doesn't allow having heterogeneous key types, when type inferring is not required (i.e: type annotations are present). So the type inferring also follows the same checker rules. e.g: Handling this case in particular: https://github.com/onflow/cadence/blob/051bf1599729d43a1e4af2d054654371739ab5a2/runtime/tests/checker/type_inference_test.go#L1136-L1146

Here, by returning an invalid type, the type-inferring says "I can't figure out a type", and then the checker says "let's ask from user".

The reason why inferring a generic type like `AnyStruct` doesn't work is because the runtime-value also requires a specific static-type/dynamic-type for runtime type checking. That runtime type is created based on the type that the checker computes.

Reply:
I see! That makes sense.

Reply:
Or are you suggesting keeping the old behavior for such cases?

Reply:
I was assuming that the common supertype of `{String: Int}` and `{Int: Int}` would be `AnyStruct`

Reply:
Right, so that pretty much what we have now. I think that's a good idea to fall back to the broader type. That makes this change also backward compatible 👍

Reply:
added in https://github.com/onflow/cadence/pull/1319/commits/09128222d8724a2db70412724912cb18d845ffd7